test/example/echo/echo.yaml

```diff
@@ -2,34 +2,54 @@
 apiVersion: v1
 kind: Service
 metadata:
-  name: echoheaders
+  name: service1
   labels:
-    app: echoheaders
+    app: service1
 spec:
   type: NodePort
   ports:
   - port: 80
     targetPort: 8080
     name: http
   selector:
-    app: echoheaders
+    app: service1
 ---
 apiVersion: extensions/v1beta1
 kind: Deployment
 metadata:
-  name: echoheaders
+  name: service1
 spec:
   replicas: 3
   template:
     metadata:
       labels:
-        app: echoheaders
+        app: service1
+      annotations:
+        pod.beta.kubernetes.io/init-containers: '[
+          {
+            "name": "iptables",
+            "image": "gcr.io/istio-test/init:test",
+            "imagePullPolicy": "Always",
+            "securityContext": {
+              "privileged": true
+            },
+            "command": ["/bin/bash", "-c", "sudo iptables -t nat -A PREROUTING -p tcp --dport 8080 -j REDIRECT --to-port 9999"]
+          }
+        ]'
     spec:
       containers:
-      - name: echoheaders
+      - name: echo
         image: gcr.io/google_containers/echoserver:1.4
         ports:
         - containerPort: 8080
+      - name: proxy
+        image: gcr.io/istio-test/runtime:test
+        args: ["proxy", "-s", "manager:8080", "--logtostderr", "-v", "2", "-p", "9999"]
+        securityContext:
+          privileged: true
+        imagePullPolicy: Always
+        ports:
+        - containerPort: 9999
```

Comment:
In our current implementation, the iptables rules are programmed by an init container which starts in privileged mode, program the rules and exits. I think that this actually doesn't need full privilege, only CAP_NET_ADMIN, so it might be possible to tighten the security a bit more.
After that, the proxy and app container starts. The proxy listens on a single port > 1024 thanks to the new changes, so it doesn't need privilege.

As for the other proxies, haproxy already knows about SO_ORIGINAL_DST so it may not need further changes (needs to be tested). Nginx has a transparent mode which might be useful and I saw a patch for SO_ORIGINAL_DST (not in the official codebase). Linkerd has published its own solution, which looks quite convoluted.
Bottom line is that other proxies need to be looked at individually to determine what is needed.
For inbound traffic other non-iptables choices are possible but for outbound I don't know about any option that does not require iptables or changes to the app.